### ISSUE TITLE
Sybil Attacker Report

### ISSUE THREAD

**Issue by rchen8** [hop-protocol/hop-airdrop#566]
# Related Addresses

```
0x97dc368deb05fe88b64d95ee15cf058852c347ac
0x2c1a420cb6c631d1d38e738d8eafa8c9dc5d0b6d
0x2dda83ea34157b05b24827a94e460d99cf9ccfb8
0xf64da7be3d8eef9e6eb2dc554253c82452e89735
0x43c200c1ea4e63ab15d1fc0439eab749c35cb1f0
0xaa4bd95b62ebf565ac4da4333f96ff3a63944c39
0x32c07bf7d22f12227c38165544f569970eebfa40
0xd5fae9a76de0bf1063eee5d4ae05be77cb08e68a
0x5f4ef5c539a56291665fcca4e6b7a6446ad11cc1
0x552cda6990392d5345d2bcd6ed59ae4acd24392c
0x79e684a001fdd917fe61bf0643d5800956c91a83
0x451a2f23b2a18f11ace266c4fe3943aea0d92cac
0x9446109c997e0ceb552be06bbd08fe8310b2fae9
0x71ff65395a8998409608f8003007c4144978169d
0x4a4e084e08e277c6e2505af3465247f58145d42c
0xe39e7f1d6b49476978775d59715a3cbdcc96e1be
0xbbd713f0de4acd0596a4a667663bfc4d1893f59c
0xa010083e4282f5a5f748c4f07833f005d1c378a4
0xf8f61e26bd65bac400abf85b827d09c5831fb8b4
0xe5dd37a8453ce27947d3f1c3586504ad5de74fff
```

# Reasoning

![Screen Shot 2022-05-21 at 10 02 20 PM](https://user-images.githubusercontent.com/10898801/169676477-b849cc0b-be85-4a7c-8d7d-72d9a44c74fd.png)

All 20 addresses belong to the same connected subgraph component, in which the edges are **Polygon** transactions between addresses. 17 of the 20 addresses form a [strongly connected component](https://en.wikipedia.org/wiki/Strongly_connected_component), which means it's very highly likely they all are controlled by the same Sybil attacker if every address is reachable from every other address (though cycles formed by both `to` and `from` transactions). The remaining three addresses `0xe5d` `0xa01` and `0xf8f` are related to the rest due to similar Hop transaction behaviors detailed in the table below, all of which on 2021-10-20.

| Date | # Addresses |
| - | - |
| 2021-10-17 | 11 |
| 2022-04-03 | 6 |
| 2021-09-28 | 5 |
| 2021-10-20 | 5 |
| 2021-10-19 | 4 |
| 2022-03-31 | 4 |
| 2021-12-26 | 4 |
| 2022-03-23 | 4 |
| 2022-03-29 | 4 |
| 2021-09-30 | 3 |
| 2021-10-16 | 3 |
| 2021-09-29 | 3 |
| 2022-04-24 | 3 |
| 2022-03-21 | 3 |
| 2022-04-04 | 3 |
| 2021-09-27 | 3 |
| 2022-03-26 | 3 |
| 2022-05-15 | 3 |
| 2022-04-25 | 3 |
| 2021-10-28 | 2 |
| 2022-04-05 | 2 |
| 2022-04-06 | 2 |
| 2021-10-22 | 2 |
| 2022-03-24 | 2 |
| 2022-04-29 | 2 |
| 2021-10-23 | 2 |

Consider 2021-10-17 as an example. All 11 addresses sent between 118-314 USDC back and forth from Gnosis and Polygon in order to fake transaction volume numbers to Sybil attack the airdrop.
```
'0x2c1', 'Gnosis', 'Polygon', '207.8585', 'USDC'
'0x2c1', 'Gnosis', 'Polygon', '209.8121', 'USDC'
'0x2dd', 'Polygon', 'Gnosis', '206.5229', 'USDC'
'0x2dd', 'Polygon', 'Gnosis', '208.2374', 'USDC'
'0xf64', 'Polygon', 'Gnosis', '209.0274', 'USDC'
'0xf64', 'Polygon', 'Gnosis', '210.9842', 'USDC'
'0x43c', 'Gnosis', 'Polygon', '222.0540', 'USDC'
'0xd5f', 'Polygon', 'Gnosis', '314.7838', 'USDC'
'0xd5f', 'Polygon', 'Gnosis', '120.3019', 'USDC'
'0x5f4', 'Gnosis', 'Polygon', '200.1465', 'USDC'
'0x5f4', 'Gnosis', 'Polygon', '236.6035', 'USDC'
'0x552', 'Gnosis', 'Polygon', '210.2029', 'USDC'
'0x552', 'Gnosis', 'Polygon', '212.1600', 'USDC'
'0x79e', 'Polygon', 'Gnosis', '194.1686', 'USDC'
'0x79e', 'Polygon', 'Gnosis', '118.1050', 'USDC'
'0x79e', 'Polygon', 'Gnosis', '118.1049', 'USDC'
'0x451', 'Gnosis', 'Polygon', '193.0139', 'USDC'
'0x451', 'Gnosis', 'Polygon', '234.0288', 'USDC'
'0x944', 'Polygon', 'Gnosis', '191.8534', 'USDC'
'0x944', 'Polygon', 'Gnosis', '232.8460', 'USDC'
'0x71f', 'Gnosis', 'Polygon', '313.5307', 'USDC'
'0x71f', 'Gnosis', 'Polygon', '119.2064', 'USDC'
```

Also consider 2021-10-20 as another example. All 5 addresses sent between 114-381 USDC or USDT back and forth between Polygon and Gnosis.
```
'0xaa4', 'Polygon', 'Gnosis', '114.4517', 'USDT'
'0xaa4', 'Gnosis', 'Polygon', '115.5515', 'USDT'
'0xaa4', 'Polygon', 'Gnosis', '116.6108', 'USDT'
'0xaa4', 'Polygon', 'Gnosis', '357.7463', 'USDC'
'0x32c', 'Gnosis', 'Polygon', '117.7071', 'USDT'
'0x32c', 'Gnosis', 'Polygon', '358.9726', 'USDC'
'0x32c', 'Polygon', 'Gnosis', '118.8006', 'USDT'
'0x32c', 'Polygon', 'Gnosis', '360.2601', 'USDC'
'0x32c', 'Gnosis', 'Polygon', '119.9095', 'USDT'
'0x32c', 'Gnosis', 'Polygon', '360.8653', 'USDC'
'0x32c', 'Polygon', 'Gnosis', '181.5771', 'USDC'
'0x32c', 'Polygon', 'Gnosis', '181.5773', 'USDC'
'0xa01', 'Polygon', 'Gnosis', '375.8700', 'USDC'
'0xa01', 'Polygon', 'Gnosis', '197.8240', 'USDC'
'0xa01', 'Polygon', 'Gnosis', '180.5359', 'USDC'
'0xf8f', 'Gnosis', 'Polygon', '133.3749', 'USDT'
'0xf8f', 'Gnosis', 'Polygon', '379.6714', 'USDC'
'0xf8f', 'Polygon', 'Gnosis', '381.0064', 'USDC'
'0xf8f', 'Gnosis', 'Polygon', '192.7985', 'USDC'
'0xf8f', 'Gnosis', 'Polygon', '190.5293', 'USDC'
'0xe5d', 'Gnosis', 'Polygon', '120.5572', 'USDT'
'0xe5d', 'Gnosis', 'Polygon', '369.9614', 'USDC'
'0xe5d', 'Polygon', 'Gnosis', '192.8576', 'USDC'
'0xe5d', 'Polygon', 'Gnosis', '179.4002', 'USDC'
'0xe5d', 'Gnosis', 'Polygon', '194.0191', 'USDC'
'0xe5d', 'Gnosis', 'Polygon', '180.5509', 'USDC'
```

# Methodology

![maxresdefault](https://user-images.githubusercontent.com/10898801/169582849-adfbfef2-c17b-4575-b3fe-8bcb8fec643c.jpeg)

I implemented the [Union-Find algorithm](https://en.wikipedia.org/wiki/Disjoint-set_data_structure), which is a famous graph algorithm that gets all of the connected subgraph components in `O(1)` time. The nodes in the graph are from the most up-to-date list of eligible airdrop addresses. The edges in the graph are from using [Covalent's API](https://www.covalenthq.com/docs/api/#/0/Get%20transactions%20for%20address/USD/1) to find transactions that connect between these addresses.

Finding the timestamps of Hop transactions per address is done using the [Hop Explorer](https://explorer.hop.exchange/) and reverse engineering their API so I can automate it. :)

# Rewards Address

0x9bb82fbf10cF4959909BAB9bE07805bd1d28D04A

**Comment by rchen8** [comment]
I realized 10 of these addresses were already included in #515 valid submission. However another 10 addresses were left out:
```
0x97dc368deb05fe88b64d95ee15cf058852c347ac
0xaa4bd95b62ebf565ac4da4333f96ff3a63944c39
0x32c07bf7d22f12227c38165544f569970eebfa40
0x9446109c997e0ceb552be06bbd08fe8310b2fae9
0x4a4e084e08e277c6e2505af3465247f58145d42c
0xe39e7f1d6b49476978775d59715a3cbdcc96e1be
0xbbd713f0de4acd0596a4a667663bfc4d1893f59c
0xa010083e4282f5a5f748c4f07833f005d1c378a4
0xf8f61e26bd65bac400abf85b827d09c5831fb8b4
0xe5dd37a8453ce27947d3f1c3586504ad5de74fff
```

**Comment by shanefontaine** [comment]
Thank you for your report @rchen8. We have verified that the addresses in this report are Sybil attackers.

The report included 10 eligible addresses as Sybil attackers which means you are eligible for 1590.211852364141779838 HOP! When Hop DAO is live, we will make a proposal for this reward — subject to a 1 year lockup, as mentioned in the [original Mirror post](https://hop.mirror.xyz/AI5fOUR0X_l0mktShDOx3mwr-hsB24gp8GvTWtS-MBc).

The qualified addresses are as follows:

```
0x97dc368deb05fe88b64d95ee15cf058852c347ac
0xaa4bd95b62ebf565ac4da4333f96ff3a63944c39
0x32c07bf7d22f12227c38165544f569970eebfa40
0x9446109c997e0ceb552be06bbd08fe8310b2fae9
0x4a4e084e08e277c6e2505af3465247f58145d42c
0xe39e7f1d6b49476978775d59715a3cbdcc96e1be
0xbbd713f0de4acd0596a4a667663bfc4d1893f59c
0xa010083e4282f5a5f748c4f07833f005d1c378a4
0xf8f61e26bd65bac400abf85b827d09c5831fb8b4
0xe5dd37a8453ce27947d3f1c3586504ad5de74fff
```
